File: apps/native/modules/visp-srt/ios/StreamState.swift
enum StreamState: String {
  case idle
  case preparing
  case connecting
  case live
  case reconnecting
  case stopping
  case error

  var shouldSuspendWhenDetached: Bool {
    switch self {
    case .idle, .preparing, .error:
      true
    case .connecting, .live, .reconnecting, .stopping:
      false
    }
  }

  func backgroundAction(pictureInPictureActiveOrStarting: Bool) -> BackgroundAction {
    switch self {
    case .connecting, .live, .reconnecting:
      pictureInPictureActiveOrStarting ? .keepStreaming : .stopStreaming
    case .idle, .preparing, .stopping, .error:
      .suspend
    }
  }
}

enum BackgroundAction {
  case keepStreaming
  case stopStreaming
  case suspend
}

File: apps/native/src/lib/stream-state.ts
import type { StreamState } from "../../modules/visp-srt";

/** Connecting, live, or reconnecting — the encoder is active. */
export function isPublishing(state: StreamState): boolean {
	switch (state) {
		case "connecting":
		case "live":
		case "reconnecting":
			return true;
		case "idle":
		case "preparing":
		case "stopping":
		case "error":
			return false;
		default: {
			const _exhaustive: never = state;
			return _exhaustive;
		}
	}
}

/** Publishing or winding down — settings stay locked and the stop control stays up. */
export function isStreamSession(state: StreamState): boolean {
	return isPublishing(state) || state === "stopping";
}
